File: web/lib/compute.ts
// TypeScript mirror of worker/compute.py, used only to generate demo-mode
// computed stats. In live mode these numbers come from the computed_stats
// table, written by the sync worker.

import type { Matchup, RosterSlot, Stat, Team } from "./types";

export const FLEX_ELIGIBLE: Record<string, string[]> = {
  FLEX: ["RB", "WR", "TE"],
  "RB/WR": ["RB", "WR"],
  "WR/TE": ["WR", "TE"],
  OP: ["QB", "RB", "WR", "TE"],
};

export function weeklyScores(
  teams: Team[],
  matchups: Matchup[],
  throughWeek: number
): Map<string, Map<number, number>> {
  const scores = new Map(teams.map((t) => [t.id, new Map<number, number>()]));
  for (const m of matchups) {
    if (!m.isFinal || m.week > throughWeek || m.isPlayoff) continue;
    scores.get(m.homeTeamId)?.set(m.week, m.homeScore);
    if (m.awayTeamId) scores.get(m.awayTeamId)?.set(m.week, m.awayScore);
  }
  return scores;
}

export function allPlayAndLuck(teams: Team[], matchups: Matchup[], throughWeek: number): Stat[] {
  const scores = weeklyScores(teams, matchups, throughWeek);
  const weeks = new Set<number>();
  scores.forEach((s) => s.forEach((_, w) => weeks.add(w)));
  const apW = new Map<string, number>();
  const apL = new Map<string, number>();
  const expected = new Map<string, number>();
  for (const w of weeks) {
    const inWeek: [string, number][] = [];
    scores.forEach((s, tid) => {
      if (s.has(w)) inWeek.push([tid, s.get(w)!]);
    });
    const n = inWeek.length;
    if (n < 2) continue;
    for (const [tid, pts] of inWeek) {
      const wins = inWeek.filter(([, o]) => o < pts).length;
      const ties = inWeek.filter(([otid, o]) => o === pts && otid !== tid).length;
      apW.set(tid, (apW.get(tid) ?? 0) + wins);
      apL.set(tid, (apL.get(tid) ?? 0) + (n - 1 - wins - ties));
      expected.set(tid, (expected.get(tid) ?? 0) + (wins + ties * 0.5) / (n - 1));
    }
  }
  const rows: Stat[] = [];
  for (const t of teams) {
    rows.push(
      { teamId: t.id, week: 0, statKey: "all_play_wins", value: apW.get(t.id) ?? 0 },
      { teamId: t.id, week: 0, statKey: "all_play_losses", value: apL.get(t.id) ?? 0 },
      {
        teamId: t.id,
        week: 0,
        statKey: "luck_delta",
        value: Math.round((t.wins + t.ties * 0.5 - (expected.get(t.id) ?? 0)) * 100) / 100,
      }
    );
    scores.get(t.id)?.forEach((s, w) =>
      rows.push({ teamId: t.id, week: w, statKey: "weekly_score", value: s })
    );
  }
  return rows;
}

export function streaks(teams: Team[], matchups: Matchup[], throughWeek: number): Stat[] {
  const results = new Map<string, [number, number][]>(teams.map((t) => [t.id, []]));
  for (const m of matchups) {
    if (!m.isFinal || m.week > throughWeek || !m.awayTeamId) continue;
    let hs = 0;
    let as = 0;
    if (m.homeScore !== m.awayScore) {
      hs = m.winnerId === m.homeTeamId ? 1 : -1;
      as = -hs;
    }
    results.get(m.homeTeamId)?.push([m.week, hs]);
    results.get(m.awayTeamId)?.push([m.week, as]);
  }
  return teams.map((t) => {
    const seq = (results.get(t.id) ?? []).sort((a, b) => a[0] - b[0]).map(([, r]) => r);
    let streak = 0;
    for (let i = seq.length - 1; i >= 0; i--) {
      const r = seq[i];
      if (r === 0 || (streak !== 0 && r > 0 !== streak > 0)) break;
      streak += r;
    }
    return { teamId: t.id, week: 0, statKey: "streak", value: streak };
  });
}

function normalize(vals: Map<string, number>): Map<string, number> {
  const arr = [...vals.values()];
  const lo = Math.min(...arr);
  const hi = Math.max(...arr);
  const out = new Map<string, number>();
  vals.forEach((v, k) => out.set(k, hi === lo ? 0.5 : (v - lo) / (hi - lo)));
  return out;
}

export function powerScores(
  teams: Team[],
  matchups: Matchup[],
  rosterSlots: RosterSlot[],
  throughWeek: number,
  prevRanks?: Map<string, number>
): Stat[] {
  const scores = weeklyScores(teams, matchups, throughWeek);
  const apStats = allPlayAndLuck(teams, matchups, throughWeek);
  const apPct = new Map<string, number>();
  for (const t of teams) {
    const w = apStats.find((s) => s.teamId === t.id && s.statKey === "all_play_wins")?.value ?? 0;
    const l = apStats.find((s) => s.teamId === t.id && s.statKey === "all_play_losses")?.value ?? 0;
    apPct.set(t.id, w + l ? w / (w + l) : 0.5);
  }
  const proj = new Map<string, number>();
  for (const rs of rosterSlots) {
    if (rs.week === throughWeek + 1 && rs.isStarter && rs.projected != null)
      proj.set(rs.teamId, (proj.get(rs.teamId) ?? 0) + rs.projected);
  }
  const recent = new Map<string, number>();
  const season = new Map<string, number>();
  for (const t of teams) {
    const vals = [...(scores.get(t.id) ?? new Map()).entries()]
      .sort((a, b) => a[0] - b[0])
      .map(([, v]) => v)
      .slice(-3);
    recent.set(t.id, vals.length ? vals.reduce((a, b) => a + b, 0) / vals.length : 0);
    season.set(t.id, t.pointsFor);
  }
  const nRecent = normalize(recent);
  const nSeason = normalize(season);
  const nProj = proj.size ? normalize(proj) : null;

  const composite = new Map<string, number>();
  for (const t of teams) {
    let v =
      0.35 * (nRecent.get(t.id) ?? 0.5) +
      0.25 * (nSeason.get(t.id) ?? 0.5) +
      0.25 * (apPct.get(t.id) ?? 0.5);
    v += 0.15 * (nProj ? nProj.get(t.id) ?? 0.5 : apPct.get(t.id) ?? 0.5);
    composite.set(t.id, Math.round(v * 1000) / 10);
  }
  const ranked = [...teams].sort((a, b) => composite.get(b.id)! - composite.get(a.id)!);
  const rows: Stat[] = [];
  ranked.forEach((t, i) => {
    rows.push(
      { teamId: t.id, week: 0, statKey: "power_score", value: composite.get(t.id)! },
      { teamId: t.id, week: 0, statKey: "power_rank", value: i + 1 }
    );
    if (prevRanks?.has(t.id))
      rows.push({ teamId: t.id, week: 0, statKey: "power_rank_prev", value: prevRanks.get(t.id)! });
  });
  return rows;
}

export function optimalLineupPoints(
  slots: { points: number; position: string }[],
  slotCounts: Record<string, number>
): number {
  const pool = [...slots].sort((a, b) => b.points - a.points);
  const used = new Set<number>();
  let total = 0;
  for (const name of ["QB", "RB", "WR", "TE", "D/ST", "K"]) {
    for (let c = 0; c < (slotCounts[name] ?? 0); c++) {
      const i = pool.findIndex((p, idx) => !used.has(idx) && p.position === name);
      if (i >= 0) {
        used.add(i);
        total += pool[i].points;
      }
    }
  }
  for (const [flexName, eligible] of Object.entries(FLEX_ELIGIBLE)) {
    for (let c = 0; c < (slotCounts[flexName] ?? 0); c++) {
      const i = pool.findIndex((p, idx) => !used.has(idx) && eligible.includes(p.position));
      if (i >= 0) {
        used.add(i);
        total += pool[i].points;
      }
    }
  }
  return Math.round(total * 100) / 100;
}

export function lineupEfficiency(
  teams: Team[],
  rosterSlots: RosterSlot[],
  positionOf: (playerId: string) => string,
  slotCounts: Record<string, number>,
  throughWeek: number
): Stat[] {
  const byTeamWeek = new Map<string, RosterSlot[]>();
  for (const rs of rosterSlots) {
    if (rs.week > throughWeek) continue;
    const key = `${rs.teamId}|${rs.week}`;
    if (!byTeamWeek.has(key)) byTeamWeek.set(key, []);
    byTeamWeek.get(key)!.push(rs);
  }
  const rows: Stat[] = [];
  const actualTotal = new Map<string, number>();
  const optimalTotal = new Map<string, number>();
  byTeamWeek.forEach((slots, key) => {
    const [tid, wStr] = key.split("|");
    const week = Number(wStr);
    const actual = slots.filter((s) => s.isStarter).reduce((a, s) => a + s.points, 0);
    const optimal = optimalLineupPoints(
      slots.map((s) => ({ points: s.points, position: positionOf(s.playerId) })),
      slotCounts
    );
    const bench = slots.filter((s) => !s.isStarter).reduce((a, s) => a + s.points, 0);
    actualTotal.set(tid, (actualTotal.get(tid) ?? 0) + actual);
    optimalTotal.set(tid, (optimalTotal.get(tid) ?? 0) + optimal);
    rows.push(
      { teamId: tid, week, statKey: "optimal_points", value: optimal },
      { teamId: tid, week, statKey: "bench_points", value: Math.round(bench * 100) / 100 }
    );
  });
  for (const t of teams) {
    const opt = optimalTotal.get(t.id) ?? 0;
    rows.push({
      teamId: t.id,
      week: 0,
      statKey: "lineup_efficiency",
      value: opt ? Math.round(((actualTotal.get(t.id) ?? 0) / opt) * 10000) / 10000 : 1,
    });
  }
  return rows;
}

export function playoffMonteCarlo(
  teams: Team[],
  matchups: Matchup[],
  throughWeek: number,
  regularSeasonWeeks: number,
  playoffTeamCount: number,
  rand: () => number,
  sims = 3000
): Stat[] {
  const scores = weeklyScores(teams, matchups, throughWeek);
  const dist = new Map<string, [number, number]>();
  for (const t of teams) {
    const vals = [...(scores.get(t.id) ?? new Map()).values()];
    const mean = vals.length ? vals.reduce((a, b) => a + b, 0) / vals.length : 100;
    const sd =
      vals.length > 1
        ? Math.sqrt(vals.reduce((a, v) => a + (v - mean) ** 2, 0) / (vals.length - 1))
        : 25;
    dist.set(t.id, [mean, Math.max(sd, 8)]);
  }
  const remaining = matchups.filter(
    (m) => !m.isPlayoff && !m.isFinal && m.week <= regularSeasonWeeks && m.awayTeamId
  );
  const byeSlots = playoffTeamCount === 6 ? 2 : 0;
  const gauss = (mean: number, sd: number) => {
    const u = Math.max(rand(), 1e-9);
    const v = rand();
    return mean + sd * Math.sqrt(-2 * Math.log(u)) * Math.cos(2 * Math.PI * v);
  };
  const made = new Map<string, number>();
  const bye = new Map<string, number>();
  const seedSum = new Map<string, number>();
  for (let s = 0; s < sims; s++) {
    const wins = new Map(teams.map((t) => [t.id, t.wins + 0.5 * t.ties]));
    const pf = new Map(teams.map((t) => [t.id, t.pointsFor]));
    for (const m of remaining) {
      const [hm, hsd] = dist.get(m.homeTeamId)!;
      const [am, asd] = dist.get(m.awayTeamId!)!;
      const hs = gauss(hm, hsd);
      const as = gauss(am, asd);
      pf.set(m.homeTeamId, pf.get(m.homeTeamId)! + hs);
      pf.set(m.awayTeamId!, pf.get(m.awayTeamId!)! + as);
      const w = hs >= as ? m.homeTeamId : m.awayTeamId!;
      wins.set(w, wins.get(w)! + 1);
    }
    const order = [...teams].sort(
      (a, b) => wins.get(b.id)! - wins.get(a.id)! || pf.get(b.id)! - pf.get(a.id)!
    );
    order.forEach((t, i) => {
      seedSum.set(t.id, (seedSum.get(t.id) ?? 0) + i + 1);
      if (i + 1 <= playoffTeamCount) made.set(t.id, (made.get(t.id) ?? 0) + 1);
      if (i + 1 <= byeSlots) bye.set(t.id, (bye.get(t.id) ?? 0) + 1);
    });
  }
  const rows: Stat[] = [];
  for (const t of teams) {
    rows.push(
      { teamId: t.id, week: 0, statKey: "playoff_odds", value: Math.round(((made.get(t.id) ?? 0) / sims) * 10000) / 10000 },
      { teamId: t.id, week: 0, statKey: "bye_odds", value: Math.round(((bye.get(t.id) ?? 0) / sims) * 10000) / 10000 },
      { teamId: t.id, week: 0, statKey: "proj_seed", value: Math.round(((seedSum.get(t.id) ?? 0) / sims) * 100) / 100 }
    );
  }
  return rows;
}

File: worker/compute.py
"""Derived metrics, computed once per sync into computed_stats.

All functions take plain dicts (rows as stored in Supabase) and return
computed_stats rows: {team_id, week, stat_key, value}. week=0 means
season-to-date.

Stat keys written here:
  all_play_wins, all_play_losses, luck_delta, streak,
  power_score, power_rank, power_rank_prev,
  optimal_points (per week), lineup_efficiency (season),
  bench_points (per week),
  playoff_odds, bye_odds, proj_seed,
  faab_spent, faab_points, weekly_score (per week)
"""

from __future__ import annotations

import random
import statistics
from collections import defaultdict

STARTER_SLOTS = {"QB", "RB", "WR", "TE", "FLEX", "RB/WR", "WR/TE", "OP", "D/ST", "K"}
FLEX_ELIGIBLE = {"FLEX": {"RB", "WR", "TE"}, "RB/WR": {"RB", "WR"},
                 "WR/TE": {"WR", "TE"}, "OP": {"QB", "RB", "WR", "TE"}}


def weekly_scores(teams: list[dict], matchups: list[dict], through_week: int) -> dict[str, dict[int, float]]:
    """team_id -> {week -> score} for completed (is_final) regular-season weeks."""
    scores: dict[str, dict[int, float]] = {t["id"]: {} for t in teams}
    for m in matchups:
        if not m["is_final"] or m["week"] > through_week or m["is_playoff"]:
            continue
        scores[m["home_team_id"]][m["week"]] = float(m["home_score"])
        if m.get("away_team_id"):
            scores[m["away_team_id"]][m["week"]] = float(m["away_score"])
    return scores


def all_play_and_luck(teams: list[dict], matchups: list[dict], through_week: int) -> list[dict]:
    """All-play record (your record if you played everyone every week) and
    luck delta (actual wins minus all-play expected wins)."""
    scores = weekly_scores(teams, matchups, through_week)
    weeks = sorted({w for s in scores.values() for w in s})
    ap_w: dict[str, int] = defaultdict(int)
    ap_l: dict[str, int] = defaultdict(int)
    expected: dict[str, float] = defaultdict(float)
    for w in weeks:
        in_week = [(tid, s[w]) for tid, s in scores.items() if w in s]
        n = len(in_week)
        if n < 2:
            continue
        for tid, pts in in_week:
            wins = sum(1 for _, other in in_week if other < pts)
            ties = sum(1 for otid, other in in_week if other == pts and otid != tid)
            ap_w[tid] += wins
            ap_l[tid] += n - 1 - wins - ties
            expected[tid] += (wins + ties * 0.5) / (n - 1)

    rows = []
    for t in teams:
        tid = t["id"]
        actual = t["wins"] + t["ties"] * 0.5
        rows += [
            {"team_id": tid, "week": 0, "stat_key": "all_play_wins", "value": ap_w[tid]},
            {"team_id": tid, "week": 0, "stat_key": "all_play_losses", "value": ap_l[tid]},
            {"team_id": tid, "week": 0, "stat_key": "luck_delta",
             "value": round(actual - expected[tid], 2)},
        ]
        for w, s in scores[tid].items():
            rows.append({"team_id": tid, "week": w, "stat_key": "weekly_score", "value": s})
    return rows


def streaks(teams: list[dict], matchups: list[dict], through_week: int) -> list[dict]:
    """Signed streak: +3 = won last 3, -2 = lost last 2."""
    results: dict[str, list[tuple[int, int]]] = defaultdict(list)  # team -> [(week, +1/-1/0)]
    for m in matchups:
        if not m["is_final"] or m["week"] > through_week or not m.get("away_team_id"):
            continue
        if m["home_score"] == m["away_score"]:
            hs = as_ = 0
        elif m["winner_id"] == m["home_team_id"]:
            hs, as_ = 1, -1
        else:
            hs, as_ = -1, 1
        results[m["home_team_id"]].append((m["week"], hs))
        results[m["away_team_id"]].append((m["week"], as_))

    rows = []
    for t in teams:
        seq = [r for _, r in sorted(results[t["id"]])]
        streak = 0
        for r in reversed(seq):
            if r == 0 or (streak != 0 and (r > 0) != (streak > 0)):
                break
            streak += r
        rows.append({"team_id": t["id"], "week": 0, "stat_key": "streak", "value": streak})
    return rows


def power_scores(teams: list[dict], matchups: list[dict], roster_slots: list[dict],
                 through_week: int, prev_ranks: dict[str, int] | None = None) -> list[dict]:
    """Composite 0-100: recent scoring (last 3 weeks, 35%), season points (25%),
    all-play win% (25%), roster strength from current projections (15%)."""
    scores = weekly_scores(teams, matchups, through_week)
    ap = {r["team_id"]: r for r in all_play_and_luck(teams, matchups, through_week)
          if r["stat_key"] == "all_play_wins"}
    apl = {r["team_id"]: r["value"] for r in all_play_and_luck(teams, matchups, through_week)
           if r["stat_key"] == "all_play_losses"}

    proj: dict[str, float] = defaultdict(float)
    cur_week = through_week + 1
    for rs in roster_slots:
        if rs["week"] == cur_week and rs["is_starter"] and rs.get("projected") is not None:
            proj[rs["team_id"]] += float(rs["projected"])

    def norm(vals: dict[str, float]) -> dict[str, float]:
        if not vals:
            return {}
        lo, hi = min(vals.values()), max(vals.values())
        if hi == lo:
            return {k: 0.5 for k in vals}
        return {k: (v - lo) / (hi - lo) for k, v in vals.items()}

    recent = {t["id"]: statistics.mean(list(dict(sorted(scores[t["id"]].items())).values())[-3:] or [0])
              for t in teams}
    season = {t["id"]: float(t["points_for"]) for t in teams}
    ap_pct = {}
    for t in teams:
        w = ap.get(t["id"], {}).get("value", 0)
        l = apl.get(t["id"], 0)
        ap_pct[t["id"]] = w / (w + l) if (w + l) else 0.5
    n_recent, n_season, n_proj = norm(recent), norm(season), norm(proj) if proj else {}

    composite = {}
    for t in teams:
        tid = t["id"]
        parts = 0.35 * n_recent.get(tid, 0.5) + 0.25 * n_season.get(tid, 0.5) + 0.25 * ap_pct[tid]
        parts += 0.15 * n_proj.get(tid, 0.5) if n_proj else 0.15 * ap_pct[tid]
        composite[tid] = round(parts * 100, 1)

    ranked = sorted(teams, key=lambda t: -composite[t["id"]])
    rows = []
    for rank, t in enumerate(ranked, 1):
        tid = t["id"]
        rows += [
            {"team_id": tid, "week": 0, "stat_key": "power_score", "value": composite[tid]},
            {"team_id": tid, "week": 0, "stat_key": "power_rank", "value": rank},
        ]
        if prev_ranks and tid in prev_ranks:
            rows.append({"team_id": tid, "week": 0, "stat_key": "power_rank_prev",
                         "value": prev_ranks[tid]})
    return rows


def optimal_lineup_points(slots: list[dict], slot_counts: dict[str, int]) -> float:
    """Best possible starter total from a week's full roster, greedy fill:
    fixed positions first (best available), then flex slots."""
    pool = sorted(slots, key=lambda s: -float(s["points"]))
    used: set[int] = set()
    total = 0.0
    for slot_name in ("QB", "RB", "WR", "TE", "D/ST", "K"):
        for _ in range(slot_counts.get(slot_name, 0)):
            for i, p in enumerate(pool):
                if i in used or p["position"] != slot_name:
                    continue
                used.add(i)
                total += float(p["points"])
                break
    for flex_name, eligible in FLEX_ELIGIBLE.items():
        for _ in range(slot_counts.get(flex_name, 0)):
            for i, p in enumerate(pool):
                if i in used or p["position"] not in eligible:
                    continue
                used.add(i)
                total += float(p["points"])
                break
    return round(total, 2)


def lineup_efficiency(teams: list[dict], roster_slots: list[dict], players_by_id: dict[str, dict],
                      slot_counts: dict[str, int], through_week: int) -> list[dict]:
    """Manager skill: actual starter points / optimal lineup points, plus weekly
    optimal_points and bench_points."""
    by_team_week: dict[tuple[str, int], list[dict]] = defaultdict(list)
    for rs in roster_slots:
        if rs["week"] > through_week:
            continue
        p = players_by_id.get(rs["player_id"], {})
        by_team_week[(rs["team_id"], rs["week"])].append({
            "points": rs["points"], "is_starter": rs["is_starter"],
            "slot": rs["slot"], "position": p.get("position", ""),
        })

    rows = []
    actual_total: dict[str, float] = defaultdict(float)
    optimal_total: dict[str, float] = defaultdict(float)
    for (tid, week), slots in by_team_week.items():
        actual = sum(float(s["points"]) for s in slots if s["is_starter"])
        optimal = optimal_lineup_points(slots, slot_counts)
        bench = sum(float(s["points"]) for s in slots if not s["is_starter"])
        actual_total[tid] += actual
        optimal_total[tid] += optimal
        rows += [
            {"team_id": tid, "week": week, "stat_key": "optimal_points", "value": optimal},
            {"team_id": tid, "week": week, "stat_key": "bench_points", "value": round(bench, 2)},
        ]
    for t in teams:
        tid = t["id"]
        eff = actual_total[tid] / optimal_total[tid] if optimal_total[tid] else 1.0
        rows.append({"team_id": tid, "week": 0, "stat_key": "lineup_efficiency",
                     "value": round(eff, 4)})
    return rows


def playoff_monte_carlo(teams: list[dict], matchups: list[dict], through_week: int,
                        regular_season_weeks: int, playoff_team_count: int,
                        sims: int = 10_000, seed: int | None = 42) -> list[dict]:
    """10k season sims from remaining schedule + each team's scoring distribution
    -> playoff %, bye %, projected seed."""
    rng = random.Random(seed)
    scores = weekly_scores(teams, matchups, through_week)
    dist = {}
    for t in teams:
        vals = list(scores[t["id"]].values())
        mean = statistics.mean(vals) if vals else 100.0
        sd = statistics.stdev(vals) if len(vals) > 1 else 25.0
        dist[t["id"]] = (mean, max(sd, 8.0))

    remaining = [m for m in matchups
                 if not m["is_playoff"] and not m["is_final"]
                 and m["week"] <= regular_season_weeks and m.get("away_team_id")]

    base_wins = {t["id"]: t["wins"] + 0.5 * t["ties"] for t in teams}
    base_pf = {t["id"]: float(t["points_for"]) for t in teams}
    bye_slots = 2 if playoff_team_count == 6 else 0

    made = defaultdict(int)
    bye = defaultdict(int)
    seed_sum = defaultdict(int)
    for _ in range(sims):
        wins = dict(base_wins)
        pf = dict(base_pf)
        for m in remaining:
            h, a = m["home_team_id"], m["away_team_id"]
            hs = rng.gauss(*dist[h])
            as_ = rng.gauss(*dist[a])
            pf[h] += hs
            pf[a] += as_
            wins[h if hs >= as_ else a] += 1
        order = sorted(teams, key=lambda t: (-wins[t["id"]], -pf[t["id"]]))
        for pos, t in enumerate(order, 1):
            seed_sum[t["id"]] += pos
            if pos <= playoff_team_count:
                made[t["id"]] += 1
            if pos <= bye_slots:
                bye[t["id"]] += 1

    rows = []
    for t in teams:
        tid = t["id"]
        rows += [
            {"team_id": tid, "week": 0, "stat_key": "playoff_odds",
             "value": round(made[tid] / sims, 4)},
            {"team_id": tid, "week": 0, "stat_key": "bye_odds",
             "value": round(bye[tid] / sims, 4)},
            {"team_id": tid, "week": 0, "stat_key": "proj_seed",
             "value": round(seed_sum[tid] / sims, 2)},
        ]
    return rows


def faab_efficiency(teams: list[dict], transactions: list[dict],
                    roster_slots: list[dict]) -> list[dict]:
    """faab_spent and faab_points (points produced by FAAB pickups for the team
    that added them, from the pickup week on)."""
    points_after: dict[tuple[str, str], float] = defaultdict(float)
    first_week: dict[tuple[str, str], int] = {}
    for tx in transactions:
        if tx["type"] not in ("WAIVER",) or not tx.get("player_in_id") or not tx.get("team_id"):
            continue
        key = (tx["team_id"], tx["player_in_id"])
        first_week[key] = min(first_week.get(key, 99), tx["week"])
    for rs in roster_slots:
        key = (rs["team_id"], rs["player_id"])
        if key in first_week and rs["week"] >= first_week[key] and rs["is_starter"]:
            points_after[key] += float(rs["points"])

    spent: dict[str, int] = defaultdict(int)
    produced: dict[str, float] = defaultdict(float)
    for tx in transactions:
        if tx["type"] == "WAIVER" and tx.get("faab_bid") and tx.get("team_id"):
            spent[tx["team_id"]] += int(tx["faab_bid"])
    for (tid, _), pts in points_after.items():
        produced[tid] += pts

    rows = []
    for t in teams:
        tid = t["id"]
        rows += [
            {"team_id": tid, "week": 0, "stat_key": "faab_spent", "value": spent[tid]},
            {"team_id": tid, "week": 0, "stat_key": "faab_points",
             "value": round(produced[tid], 2)},
        ]
    return rows
